doc/settings.rst
.. _cli:
.. py:module:: ui

Command Line Interface
======================

Introduction
------------

The file ``settings.py`` contains the configuration of BATMAN. It can be devided into 2 mandatory blocks and 3 optionnal block. Fields in brackets are optionnal and there is no specific order to respect.

.. note:: A prefilled example is shown in ``settings.json`` located in ``test_cases/Snippets``.

Help of the CLI can be triggered with::
    
    batman -h

    usage: BATMAN [-h] [--version] [-v] [-c] [-s] [-o OUTPUT] [-r] [-n] [-u] [-q]
              settings

    BATMAN creates a surrogate model and perform UQ.

    positional arguments:
      settings              path to settings file

    optional arguments:
      -h, --help            show this help message and exit
      --version             show program's version number and exit
      -v, --verbose         set verbosity from WARNING to DEBUG, [default: False]
      -c, --check           check settings, [default: False]
      -s, --save-snapshots  save the snapshots to disk when using a function,
                            [default: False]
      -o OUTPUT, --output OUTPUT
                            path to output directory, [default: ./output]
      -r, --restart         restart pod, [default: False]
      -n, --no-surrogate    do not compute surrogate but read it from disk,
                            [default: False]
      -u, --uq              Uncertainty Quantification study, [default: False].
      -q, --q2              estimate Q2 and find the point with max MSE, [default:
                            False]    

.. note:: Fields in square brackets are optionnals.

Block 1 - Space of Parameters
-----------------------------

First of all, we define the parameter space using an hypercube. Taking the minimal and the maximal value along all coordinates allow to describe it.

.. figure:: fig/hypercube.pdf

   3-dimentionnal hypercube

.. code-block:: python

    "space": {
        "corners": [
            [15.0, 2500.0],
            [60.0, 6000.0]
        ],
        "sampling": {
            "init_size": 4,
            "method": "halton"
        },
        "resampling":{
            "delta_space": 0.08,
            "resamp_size": 0,
            "method": "sigma",
            "hybrid": [["sigma", 4], ["loo_sobol", 2]],
            "q2_criteria": 0.9
        }
    }

+ ``corners``: define the space using the two corners of the hypercube ``[[min], [max]]``,
+ ``sampling``: define the configuration of the sample. This can either be; a list of sample
  as an array_like of shape (n_samples, n_features); or a dictionary with
  the following:

    * ``init_size``: define the initial number of snapshots,
    * ``method``: method to create the DoE, can be *uniform*, *faure*, *halton*,
      *sobol*, *sobolscramble*, *lhs* (Latin Hypercube Sampling), *lhsc* (Latin Hypercube  Sampling Centered) or *lhsopt* (optimized LHS), *saltelli*,
    * [``distributions``]: A list of distributions. Ex for two input variables:
      ``["Uniform(15., 60.)", "Normal(4035., 400.)"]``.

+ [``resampling``]: to do resampling, fill this dictionary

    * ``delta_space``: the percentage of space to shrink to not resample close to boundaries. For ``0.08``,
      the available space for resampling will be shrinked by 8%.
    * ``resamp_size``: number of point to add in the parameter space.
    * ``method``: to be choosen from ``sigma``, ``loo_sigma``, ``loo_sobol``, ``hybrid``, ``discrepancy``, ``optimization``, ``extrema``.
    * [``hybrid``]: if method is ``hybrid``. You have to define a generator which is a list
      ``[["method", n_snapshot]]``
    * ``q2_criteria``: stopping criterion based on the quality estimation of the model.

The method used to create the DoE is paramount. It ensures that that the physics
will be captured correclty all over the domain of interest, see :ref:`Space <space>`.
All *faure*, *halton* and *sobol* methods are low discrepancy sequences with
good filling properties. *saltelli* is particular as it will create a DoE for
the computation of *Sobol'* indices using *Saltelli*'s formulation.

When *distribution* is set, a join distribution is built an is used to perform
an inverse transformation (inverse CDF) on the sample. This allows to have a
low discrepancy sample will still following some distribution.

Regarding the resampling, all methods need a good initial sample. Meanning that the quality is about :math:`Q_2\sim0.5`. ``loo_sigma, loo_sobol`` work better than ``sigma`` in high dimentionnal cases (>2).

.. warning:: If using a PC surrogate model, the only possibilities are ``discrepancy`` and ``extrema``. Furthermore, sampling ``method`` must be set as a list of distributions.

Block 2 - Snapshot provider
---------------------------

A snapshot defines a simulation.

.. code-block:: python

    "snapshot": {
        "max_workers": 5,
        "io": {
            "shapes": {
                "0": [
                    [400]
                ]
            },
            "format": "fmt_tp_fortran",
            "variables": ["X", "F"],
            "point_filename": "point.json",
            "filenames": {
                "0": ["function.dat"]
            },
            "template_directory": "output/snapshots/0/batman-data/point.json",
            "parameter_names": ["x1", "x2"]
        },
        "provider": {
            "command": "bash",
            "timeout": 20,
            "context": "data",
            "script": "data/script.sh",
            "clean": false,
            "private-directory": "batman-data",
            "data-directory": "cfd-output-data",
            "restart": "False"
        }

+ ``max_workers``: maximum number of simultaneous running snapshot,
+ ``shapes``: shape per variable and per file,
+ ``format``:  ``npz``, ``fmt_tp_fortran`` (BATMAN) or all Antares formats if installed,
+ ``variables``: names of the variables to treat and contained in a snapshot,
+ ``point_filename``: name of the file that contains the coordinates of a point in the parameter space,
+ ``filenames``: dictionary of list, if several filenames, several independant analysis are done. It contains the output ``variables`` of a snapshot to use by BATMAN,
+ ``template_directory``: path to the point file name of the first snapshot, used to restart from existing jobs not created with BATMAN,
+ ``parameter_names``: names of the parameters.

The ``provider`` defines what is a simulation. If we simply want to evaluate a python function, we can pass the ``function_name`` as a string. ``function_name.py`` will be imported and the function named ``f`` will be used. Otherwize, for a complexe case, use a dictionary:

+ ``command``: command to use to launch the script,
+ ``timeout``: timeout of each jobs in seconds,
+ ``context``: directiory where the data for the BATMAN computations are stored,
+ ``script``: path of the script or batch to run,
+ ``clean``: delete after run all except what is inside ``private-directory``
+ ``private-directory``: folder containing the ``point_filename`` and the result of the snapshot
+ ``data-directory``: output folder to store the ``filenames``,
+ ``restart``: restart the computation if the job has failed.

Optionnal Block 3 - Surrogate
-----------------------------

Set up the surrogate model strategy to use. See :ref:`Surrogate <surrogate>`.

.. code-block:: python

    "prediction": {
        "method": "kriging",
        "predictions": [[30, 4000], [35, 3550]]
    }

+ ``method``: method used to generate a snapshot one of *rbf* (Radial Basic Function), *kriging*, *pc* (polynomial chaos expension) or *evofusion* method.
+ [``predictions``]: set of points to predict.

For *kriging* the following extra attributes **can** be set: 

+ [``kernel``]: kernel to use. Ex: ``"ConstantKernel() + Matern(length_scale=1., nu=1.5)"``.
+ [``noise``]: noise level as boolean or as a float.

For *pc* the following extra attributes **must** be set: 

+ ``strategy``: either using quadrature or least square one of *Quad* or *LS*.
+ ``degree``: the polynomial degree.

.. note:: When using *pc*, the ``sampling`` must be set to a list of distributions.

For *evofusion* the following extra attributes **must** be set: 

+ ``cost_ratio``: cost ratio in terms of function evaluation between high and low fidelity models.
+ ``grand_cost``: total cost of the study in terms of number of function evaluation of the high fidelity model.

.. note:: We can fill *directly* the number of points into the brackets or *indirectly* using the script ``prediction.py`` located in ``test_cases/Snippets``.



Optionnal Block 4 - UQ
----------------------

Uncertainty Quantification (UQ), see :ref:`UQ <uq>`.

.. code-block:: python

    "uq": {
        "test": "Channel_Flow"
        "sample": 1000,
        "method": "sobol"
        "pdf": ["Normal(4035., 400)", "Uniform(15, 60)"],
        "type": "aggregated",
    }

+ ``test``: use a test method for indices comparison and quality calculation. Use one of: *Rosenbrock*, *Michalewicz*, *Ishigami*, *G_Function*, *Channel_Flow*,
+ ``sample``: number of points per sample to use for SA,
+ ``method``: type of Sobol analysis: *sobol*, *FAST* (Fourier Amplitude Sensitivity Testing). If FAST, no second-order indices are computed and defining a surrogate model is mandatory.
+ ``type``: type of indices: *aggregated* or *block*.

+ ``pdf`` *Probability density function* for uncertainty propagation. Enter the PDF of the inputs,
  as list of openturns distributions. Ex: x1-Normal(mu, sigma), x2-Uniform(inf, sup)
  => ``["Uniform(15., 60.)", "Normal(4035., 400.)"]``


Optionnal Block 5 - POD
-----------------------

POD (or Proper Orthogonal Decomposition) is a approach to help reduce amount of data.

.. code-block:: python

     "pod": {
        "dim_max": 100,
        "tolerance": 0.99,
        "type": "static"
     }

+ ``tolerance``: tolerance of the modes to be kept. A percentage of the sum of the singular values, values that account for less than this tolerance are ignored,
+ ``dim_max``: maximum number of modes to be kept,
+ ``type``: type of POD to perform: *static* or *dynamic*.

The dynamic POD allows to update the POD once a snapshot is availlable. Hence a POD can be restarted when doing resampling for example.


Optionnal Block 6 - Visualization
---------------------------------

Set up for the visualization options. Batman creates a response function (1 input parameter), response surfaces (2 to 4 input parameters) or a Kiviat graph (more than 4 input parameters). See :ref:`Visualization <visualization>`.

.. code-block:: python

     "visualization": {
        "doe": true,
        "resampling": true,
        "axis_disc": [20, 20],
        "flabel": "Cost function",
        "plabels": ["X", "Y"],
        "feat_order": [1, 2],
        "ticks_nbr": 14,
        "range_cbar": [0.0, 2.3],
        "contours": [0.5, 1.0, 1.5],
        "kiviat_fill": true
     }

+ ``doe``: Boolean. If *true*, the Design of Experiment is represented on the response surface by black dots. Defaults value is *false*,
+ ``resampling``: Boolean. If *true*, Design of Experiment corresponding to the resampling points are displayed in a different color. Such points are represented by red triangles. Only activates if doe is *true*,
+ ``axis_disc``: Integers. Discretisation of each axis. Indicated value for the x and the y axis modify the surface resolution, while values corresponding the the 3rd and 4th parameters impact the frame number per movie and the movie number,
+ ``flabel``: String. Name of the cost function,
+ ``plabels``: Strings. Name of the input parameters to be plotted on each axis,
+ ``feat_order``: Integers. Associate each input parameter to an axis, the first indicated number corresponding to the parameter to be plotted on the x-axis, etc... A size equal to the input parameter number is expected, all integers from 1 to the parameter number should be used. Default is *[1, 2, 3, 4]*,
+ ``ticks_nbr``: Integer. Number of ticks on the colorbar (Display n-1 colors). Default is *10*,
+ ``range_cbar``: Floats. Minimum and maximum values on the colorbar,
+ ``contours``: Floats. Values of the iso-contours to be plotted on the response surface,
+ ``kiviat_fill``: Boolean. If *true*, will fill the surface of the Kiviat plot.


.. py:module:: driver
.. py:currentmodule:: driver

Driver module
-------------

.. automodule:: batman.driver
   :members:
   :undoc-members:


CHANGELOG.rst
.. _changes:

===============
Release History
===============

Version 1.7 - Lucius (under development)
========================================

New features
------------

    - Add :func:`space.Space.discrepancy` function in ``Space``,
    - Refactor :class:`space.Space`, :class:`uq.UQ`,
      :class:`space.Refiner` initiate without dictionnary,
    - Refactor :class:`surrogate.PC` and add options in
      settings: ``degree``, ``strategy`` (*Quad* or *LS*),
    - Add :func:`space.Refiner.discrepancy`, and
      :func:`space.Refiner.sigma_discrepancy`
    - Add quality for every surrogate model,
    - Be able to bypass POD and surrogate in settings,
    - Surrogate facultative for UQ,
    - Add :mod:`visualization` with: Kiviat, DoE, HDR and response_surface,
      response_surface *by Robin Campet*,
    - Add optimized LHS, *by Vincent Baudoui*,
    - Concurrent CI, *by Cyril Fournier*,
    - pylint/pycodestyle for CI and Python2 on develop and master branches,
    - Add *about* section in doc.

Enhancements
------------

    - Remove loops in predictors, ``zip``,
    - Backend overwright for matplotlib removed,
    - Remove ``otwrapy``,
    - JSON schema constrained for surrogate and sampling,
    - Remove support for OpenTURNS < 1.8,
    - Add some options for :class:`functions.MascaretApi`,
    - Coverage and tests raised to 90%.

Bug fixes
---------

    - Quality with multimodes with POD,
    - List in sampling settings,
    - Restart and restart from files,
    - Variance and FAST,
    - Double prompt in python 2.7, *by Vincent Baudoui*,
    - Inputs mocking in tests,
    - DoE diagonal scaling,
    - :class:`functions.MascaretApi` ``multi_eval``,
    - Block indices,
    - Normalize output in :class:`surrogate.Kriging`.


Version 1.6 - Selina
====================

New features
------------

    - Add :class:`functions.MascaretApi`,
    - Add *Evofusion* with :class:`surrogate.Evofusion`,
    - Add *Expected Improvement* with :func:`space.Refiner.optimization`,
    - Be able to have a discrete parameter.

Enhancements
------------

    - Allow ``*args`` and ``**kwargs`` in ``@multi_eval``,
    - Add some analytical functions for optimization and multifidelity tests,
    - Do not use anymore ``.size`` for space sizing,
    - Add test for DoE,
    - Add PDFs of references to doc,
    - Refinements methods work with discrete values using an optimizer decorator,
    - Changed some loops in favor of list comprehensions,
    - Clean UI by removing prediction option,
    - Remove MPI dependencie.

Bug fixes
---------

    - Sensitivity indices with n-dimensional output changing ``Martinez``,
    - A copy of the space is done for scaled points for surrogate fitting,
    - Uniform sampling was not set properly,
    - Backend for ``matplotlib`` is now properly switched,
    - POD quality was not computed in case of varying number of modes.


Version 1.5 - Oswald
====================

New features
------------

    - Python3 support,
    - Add :class:`surrogate.surrogate_model.SurrogateModel`,
    - Add progress bar during quality computation,
    - Use pathos for multiprocessing during LOO and Kriging.
      New :class:`misc.nested_pool` allow nested pool.
    - Unittests and functionnal tests using Pytest,
    - Antares wrapper used for IO,
    - OT1.8 support and use of new SA classes,
    - Add plot of aggregated indices,
    - Add *snipets*,
    - Add correlation and covariance matrices,
    - Add DoE visualization in *n-dimension*,
    - Hypercube for refinement created using discrete and global optimization,
    - Merge some ``PyUQ`` functions and add :class:`surrogate.PC`.
    
Enhancements
------------

    - Refactor :mod:`space`, :mod:`surrogate`, :mod:`snapshots`, :mod:`pod`,
    - Rewrite ``settings.json``,
    - POD is now optional,
    - Use a wrapper for OT evaluations with ``otwrapy``,
    - Comment capability to ``settings.json``,
    - Doc cleanning,
    - Use :mod:`functions` to test model error,
    - Remove some MPI functions,
    - Simplify hybrid navigator using generator.

Bug fixes
---------

    - Use of timeout option,
    - Remove ``snapshots.tar``,
    - FAST indices for aggregated indices,
    - Update keyword for POD,
    - Verbosity with quality,
    - Setup dependencies,
    - Some RBF cleanning,
    - Term *MSE* changed to *sigma*,
    - Snapshot ``repr``,
    - Add *.so* when packaging.


Version 1.4
===========

New features
------------

    - Enhance :class:`surrogate.kriging`: adimentionize input parameters,
      use anisotropic kernel and use genetic algorithm for parameters optimization
    - Settings are now written in JSON and checked using a schema
    - Ask for confirmation of output if exists: if no, ask for restarting from files
    - Add post-treatment example scripts in ``test_cases/Post-treatment``

Enhancements
------------

    - Save points of the DOE as human readable file
    - Add branch and commit information to log
    - Add doc for tutorial, space, surrogate and pod
    - Change Scikit-Learn to stable 0.18
    - Restart option ``-r`` now working properly
    - Create a :mod:`misc` which contains logging and json schema

Bug fixes
---------

    - Refiner navigator loops correctly
    - LOOCV working for multimodes
    - Revert Q2 variance to use ``eval_ref``
    - Avoid extra POD quality when using LOOCV strategies
    - Popping space was not working properly


Version 1.3
===========

New features
------------

    - Add resampling strategies with :class:`space.refiner`. Possibilities are:
      ``None, MSE, loo_mse, loo_sobol, hybrid``
    - Computation of the error of the pod *Q2* with option ``-q2``. Uses *Kriging*
    - Aggregated and block *Sobol'* indices are computed using a set of keywords:
      ``aggregated`` and ``block``
    - Add the possibility to chose the *PDF* for propagation. (settings)
    - *Sobol'* map are computed using the keyword ``aggregated``
    - A *Sphinx* documentation is available in: ``/doc``


test_cases/Ishigami/settings.py
# parameters space
space = {
# Lower and upper end points (corners) that define a portion of space.
# format : 2-tuple of tuples with end points coordinates.
   'corners'         : ((-3.1415, -3.1415, -3.1415),
                        (3.1415, 3.1415, 3.1415),),
   'delta_space'     : 0.01,                         
# Maximum number of point, used for pod automatic resampling
# format : integer
    'size_max' : 7 ,
# Points provider
# Could be a list of points or a dictionary with sampling parameters
    'provider' : {
    # Method used to generate the points
    # format : one of 'uniform', 'halton', 'sobol', 'lhcc', 'lhcr'
        'method' : 'halton',
    # Number of samples to be generated
    # format : integer
        'size' : 6 ,
    }
}

# import functions

snapshot = {
# Snapshot provider
# Maximum number of simultaneous running snapshot provider
    'max_workers' : 50,
# Input output settings
    'io' : {
    # Names of the parameters
    # format : list of strings
        'parameter_names': ['x1','x2', 'x3'],
    # File format
    # format : one of 'fmt_tp', 'numpy'
        'format' : 'fmt_tp',
    # File names for each mpi cpu
    # When ran on only 1 cpu, all filenames are gathered
    # format : list of strings
        'filenames' : {0: ['function.dat']},
    # Name of the file that contains the coordinates
    # of a point in the space of parameters
        'point_filename' : 'header.py',
    # Directory to store io templates
        'template_directory' : None,
    # Names of the variables contained in a snapshot
    # format : list of strings
        'variables' : ['F'],
    # Shapes of one variable for each file and each mpi cpu
    # When ran on only 1 cpu, all shapes are gathered
        'shapes' : {0: [(1,)]},
    },
}


pod = {
# Tolerance of the modes to be kept.
# A percentage of the sum of the singular values, values that account for less than of this tolerance are ignored.
# format : float
    'tolerance' : 0.99,
# Maximum number of modes to be kept
# format : integer
    'dim_max'   : 100,
# Type of pod to perform.
# format : one of 'static', 'dynamic'
    'type'      : 'static',
# Resampling strategy: None, 'MSE', 'loo_mse', 'loo_sobol', 'extrema', 'hybrid'
    'resample'  : 'MSE',
    'strategy' : (('MSE', 2), ('loo_sobol', 0), ('extrema', 1)),
# Stopping criterion for resampling
# format : float
    'quality'   : 0.80,            
# Server settings
# None means no server, the pod processing is run from the main python interpreter
    'server' : None,
# Otherwise the pod processing is run in another process or on another computer, the settings is a dictionary with the following parameters.
    # 'server' : {
    #     # Server hostname with port
    #         'port'   : 800199,
    #     # Python executable that runs the server
    #         'python' : 'python',
    #     },
}

prediction = {
# Method used to generate a snapshot
# format : one of 'rbf' , 'kriging'
    'method' : 'kriging',
# Set of points at which the predictions are made
# format : list of tuples of floats
    'points' : [],
}

uq = {
# Type of Sobol analysis: 'sobol', 'FAST' (if FAST, no second-order indices)
    'method' : 'sobol',
# Type of indices we want: 'aggregated', 'block'
    'type' : 'aggregated',
# Use a test method: 'Ishigami'
    'test' : 'Ishigami',
    'sample' : 1000 ,
# Uncertainty propagation. Enter the PDF of the inputs. x1: Normal(mu, sigma), x2: Uniform(inf, sup)
    'pdf' : ['Uniform(-3.1415, 3.1415)', 'Uniform(-3.1415, 3.1415)', 'Uniform(-3.1415, 3.1415)']
}

import numpy as np
import itertools
num = 25
x = np.linspace(space['corners'][0][0], space['corners'][1][0], num=num)
y = np.linspace(space['corners'][0][1], space['corners'][1][1], num=num)
z = np.linspace(space['corners'][0][1], space['corners'][1][1], num=num)
xyz = []
for i, j, k in itertools.product(x, y, z):
    xyz += [(float(i), float(j), float(k))]
    prediction['points'] = xyz
